source/slang/slang-emit-source-writer.cpp

```diff
@@ -239,35 +239,197 @@ void SourceWriter::emit(double value)
     // 3) We need to be sure that floating-point literals specified by the user will
     //    "round-trip" and turn into the same value when parsed back in. This means
     //    that we need to print a reasonable number of digits of precision.
-    //
-    // For right now, the easiest option that can balance these is to use
-    // the C++ standard library `iostream`s, because they support an explicit locale,
-    // and can (hopefully) print floating-point numbers accurately.
-    //
-    // Eventually, the right move here would be to implement proper floating-point
-    // number formatting ourselves, but that would require extensive testing to
-    // make sure we get it right.
-
-    std::ostringstream stream;
-    stream.imbue(std::locale::classic());
-    stream.setf(std::ios::fixed, std::ios::floatfield);
-    stream.precision(20);
-    stream << value;
-    auto str = stream.str();
-    auto slice = UnownedStringSlice(str.c_str());
-    // Remove redundant trailing 0s.
-    if (slice.end() > slice.begin())
+
+    char result[30];
+    char *next = result;
+    double valueAbs = value;
+
+    // Handle negative sign
+    const long long* d2l = reinterpret_cast<long long*>(&value);
+    const bool negative = (*d2l < 0);
+    if (negative)
+    {
+        *next++ = '-';
+        valueAbs = -valueAbs;
+    }
+
+    // Handle one of five cases
+    if (std::isnan(value))
+    {
+        *next++ = 'n';
+        *next++ = 'a';
+        *next++ = 'n';
+    }
+    else if (std::isinf(value))
     {
-        auto lastChar = slice.end() - 1;
-        while (lastChar > slice.begin() && *lastChar == '0')
-            lastChar--;
-        if (*lastChar == '.')
-            lastChar++;
-        if (lastChar > slice.end() - 1)
-            lastChar = slice.end() - 1;
-        slice = slice.subString(0, lastChar - slice.begin() + 1);
+        *next++ = 'i';
+        *next++ = 'n';
+        *next++ = 'f';
     }
-    emit(slice);
+    else if (value == 0)
+    {
+        *next++ = '0';
+        *next++ = '.';
+        *next++ = '0';
+    }
+    else
+    {
+        int exponent = 0;
+        double mantissa = valueAbs;
+        const double rounding = std::pow(10, -maxFracDigits) / 2;
+
+        // Normalize the mantissa
+        while (mantissa >= 10.0) {
```

Comment:
Instead of implementing all these logic ourselves, can we call `ostream << std::scientific << std::setprecision(precision) << value` to emit a float with specific number of digits in scientific notation?

Reply:
Currently the implementation uses "fixed" format and if I switch to "scientific" format, it breaks a few existing tests.
And simple values will become harder to read. As an example, "1.0" with precision value 20 will become, "1.00000000000000000000e+00".
To void the problem, I will have to check the exponent value and decide if the value is better with scientific format or fixed format. In other words, half of my current implementation will still be there.

Note that there was a previous comment saying,
"// Eventually, the right move here would be to implement proper floating-point number formatting ourselves, "
And I agree to this statement in three reasons.
1. we will have more explicit control over how floating values are emitted.
2. our coding guide line stated not to use iostream.
3. the performance will be better because each call to the member function of ostringstream will internally take a mutex lock to prevent the system wide locale change in the middle of text parsing. I have seen from my previous project that this mutex locking cost almost 30% of the time for processing the text messages.

Reply:
How about this:
1. use isnan / isinf to check for nan and infs, and early out those cases.
2. use frexp to extract exponent bits.
3. if exponent is less than a small value, use fixed output. truncate 0s in the end.
4. otherwise, use scientific output, truncate 0s before "e".

I want to avoid implementing our own logic as much as possible, because unless we also roll an extensive test suite on this, we can't guarantee its correctness.

Reply:
I will change to your suggestions.